### PR TITLE
Improve the error message when a PR lacks the necessary approvals

### PR DESCRIPTION
**Pull request by huydhn**
The error message from https://github.com/pytorch/pytorch/pull/115329#issuecomment-1857135047 is pretty confusing because it lists some random `pytorch/metamates` folks from `superuser` merge rule.  My attempt here is to make the error message clearer by pointing out:

* All the matching merge rules and
* Their list of approvers

The message will now become:

```
Approvers from one of the follow rules are needed:
- Core Reviewers (1, 2, 3, 4, 5, ...)
- Core Maintainers (1, 2)
```

### CODE REVIEW

**Review by malfet — APPROVED** [review]

**Comment by malfet** [comment]
Looks good, but lets see how it works in practice

**Review by PaliC — APPROVED** [review]

**Review by atalman — APPROVED** [review]

File: .github/scripts/trymerge.py [hunk 1]
```diff
@@ -1331,24 +1334,33 @@ def find_matching_merge_rule(
...
+
+                all_rule_approvers[rule.name] = rule.approved_by
+                # Prepare the reject reason
+                all_rule_approvers_msg = [
+                    f"- {name} ({', '.join(approved_by[:5])}{', ...' if len(approved_by) > 5 else ''})"
+                    for name, approved_by in all_rule_approvers.items()
+                ]
+
+                reject_reason = (
+                    "Approvers from one of the following rules are needed:\n"
```
Comment (ZainRizvi) [inline .github/scripts/trymerge.py:1360]: Looking at it from the user's perspective, they might not see any rules

```suggestion
                    "Approvers from one of the following sets are needed:\n"
```

**Review by ZainRizvi — APPROVED** [review]
This was always really confusing. This is a nice approach to fixing it, thanks!

**Pull request closed without merging**